File: Deploy/frontend/static/js/pages/face_search.js
/* Face Search and Verify page */

function renderFaceSearch() {
  setContent(`
    <div class="tabs" id="fs-tabs">
      <div class="tab active" data-tab="search">Search (1-to-N)</div>
      <div class="tab" data-tab="verify">Verify (1-to-1)</div>
    </div>

    <!-- Search tab -->
    <div class="tab-content active" id="tc-search">
      <div class="two-col">
        <div>
          <div class="card">
            <div class="card-title">Query Image</div>
            <div class="upload-zone">
              <input type="file" id="search-file" accept="image/*" onchange="_onSearchFileChange()" />
              <div class="upload-icon-placeholder"></div>
              <div class="upload-text"><strong>Choose image</strong> to search against</div>
            </div>
            <img id="search-preview" style="display:none;margin-top:12px;border-radius:8px;max-height:280px;width:100%;object-fit:contain" />
            <div class="form-group" style="margin-top:14px">
              <label class="form-label">Top K</label>
              <div style="display:flex;align-items:center;gap:12px">
                <input type="range" id="search-topk" min="1" max="10" value="5"
                       oninput="el('topk-val').textContent=this.value"
                       style="flex:1;cursor:pointer" />
                <span id="topk-val" style="color:var(--blue);font-weight:700;min-width:20px">5</span>
              </div>
            </div>
            <button class="btn btn-primary" onclick="_doSearch()">Search</button>
          </div>
        </div>
        <div>
          <div class="card">
            <div class="card-title">Top-K Results</div>
            <div id="search-results">
              <div class="empty-state"><p>Upload an image to search</p></div>
            </div>
          </div>
        </div>
      </div>
    </div>

    <!-- Verify tab -->
    <div class="tab-content" id="tc-verify">
      <div class="two-col">
        <div>
          <div class="card">
            <div class="card-title">Verification</div>
            <div class="form-group">
              <label class="form-label">Select person *</label>
              <select class="form-select" id="verify-person"></select>
            </div>
            <div class="form-group">
              <label class="form-label">Query image</label>
              <div class="upload-zone">
                <input type="file" id="verify-file" accept="image/*" onchange="_onVerifyFileChange()" />
                <div class="upload-icon-placeholder"></div>
                <div class="upload-text"><strong>Choose image</strong> to verify</div>
              </div>
              <img id="verify-preview" style="display:none;margin-top:12px;border-radius:8px;max-height:280px;width:100%;object-fit:contain" />
            </div>
            <button class="btn btn-primary" onclick="_doVerify()">Verify</button>
          </div>
        </div>
        <div>
          <div class="card">
            <div class="card-title">Result</div>
            <div id="verify-results">
              <div class="empty-state"><p>Select a person and upload an image</p></div>
            </div>
          </div>
        </div>
      </div>
    </div>
  `);

  // Tabs
  const tabs = qsa('.tab', el('fs-tabs'));
  tabs.forEach(tab => {
    tab.addEventListener('click', () => {
      tabs.forEach(t => t.classList.remove('active'));
      qsa('.tab-content').forEach(c => c.classList.remove('active'));
      tab.classList.add('active');
      el('tc-' + tab.dataset.tab).classList.add('active');
    });
  });

  _loadVerifyPeople();
}

function _onSearchFileChange() {
  const f = el('search-file').files[0]; if (!f) return;
  const p = el('search-preview'); p.src = URL.createObjectURL(f); p.style.display = 'block';
}

function _onVerifyFileChange() {
  const f = el('verify-file').files[0]; if (!f) return;
  const p = el('verify-preview'); p.src = URL.createObjectURL(f); p.style.display = 'block';
}

async function _doSearch() {
  const fileInput = el('search-file');
  if (!fileInput.files[0]) { Toast.warning('Select an image first'); return; }

  const topk   = el('search-topk').value;
  const fd     = new FormData();
  fd.append('image', fileInput.files[0]);

  const resEl  = el('search-results');
  resEl.innerHTML = '<div class="loading-spinner"><div class="spinner"></div></div>';

  try {
    const data      = await API.post(`/face/search?top_k=${topk}`, fd);
    const results   = data.results || [];
    const threshold = 0.35;

    if (!results.length) {
      resEl.innerHTML = '<div class="empty-state"><p>No results found</p></div>';
      return;
    }

    const best    = results[0];
    const isMatch = best.similarity >= threshold;

    resEl.innerHTML = `
      <div style="margin-bottom:16px;padding:14px;border-radius:8px;
           background:${isMatch ? 'var(--green-bg)' : 'var(--red-bg)'};
           border:1px solid ${isMatch ? 'var(--green)' : 'var(--red)'}">
        <div style="font-size:15px;font-weight:700;color:${isMatch ? 'var(--green)' : 'var(--red)'}">
          ${isMatch ? 'MATCH' : 'NO MATCH / UNKNOWN'}
        </div>
        <div style="font-size:13px;margin-top:4px;color:var(--text-muted)">
          Best: <strong style="color:var(--text)">${best.full_name || best.person_id}</strong>
          &nbsp;&bull;&nbsp; similarity = ${Number(best.similarity).toFixed(4)}
          &nbsp;&bull;&nbsp; threshold = ${threshold}
        </div>
      </div>

      <p style="font-size:11px;color:var(--text-muted);margin-bottom:10px;text-transform:uppercase;
                font-weight:600;letter-spacing:.05em">
        ${data.num_faces || 0} face(s) detected &nbsp;&bull;&nbsp; Top ${results.length} results
      </p>

      <div class="table-wrap">
        <table>
          <thead><tr>
            <th>#</th><th>Person ID</th><th>Full Name</th><th>Code</th><th>Class</th>
            <th>Status</th><th>Similarity</th>
          </tr></thead>
          <tbody>${results.map((r, i) => `
            <tr style="${r.similarity >= threshold ? 'background:rgba(63,185,80,.05)' : ''}">
              <td class="td-muted">${i+1}</td>
              <td class="td-mono">${r.person_id}</td>
              <td class="td-strong">${r.full_name || '&ndash;'}</td>
              <td class="td-muted">${r.student_code || '&ndash;'}</td>
              <td class="td-muted">${r.class_name || '&ndash;'}</td>
              <td>${r.status ? fmt.badge(r.status) : '&ndash;'}</td>
              <td>${fmt.simBar(r.similarity, threshold)}</td>
            </tr>`).join('')}
          </tbody>
        </table>
      </div>`;
  } catch(e) {
    Toast.error(e.message);
    resEl.innerHTML = `<p style="color:var(--red)">${e.message}</p>`;
  }
}

async function _loadVerifyPeople() {
  try {
    const people = await API.get('/people');
    const sel    = el('verify-person');
    if (!sel) return;
    sel.innerHTML = people.length
      ? people.map(p => `<option value="${p.person_id}">${p.person_id} — ${p.full_name || ''}</option>`).join('')
      : '<option>No people found</option>';
  } catch {}
}

async function _doVerify() {
  const pid  = el('verify-person')?.value;
  const file = el('verify-file')?.files[0];
  if (!pid)  { Toast.warning('Select a person first'); return; }
  if (!file) { Toast.warning('Select an image first'); return; }

  const resEl = el('verify-results');
  resEl.innerHTML = '<div class="loading-spinner"><div class="spinner"></div></div>';

  const fd = new FormData();
  fd.append('person_id', pid);
  fd.append('image', file);

  try {
    const r        = await API.post('/face/verify', fd);
    const verified = r.is_verified;
    const sim      = Number(r.similarity || 0);

    resEl.innerHTML = `
      <div style="text-align:center;padding:28px 0">
        <div style="font-size:13px;font-weight:700;letter-spacing:.08em;
             color:${verified ? 'var(--green)' : 'var(--red)'};margin-bottom:8px">
          ${verified ? 'VERIFIED' : 'NOT VERIFIED'}
        </div>
        <div style="display:inline-block;width:80px;height:80px;border-radius:50%;
             border:3px solid ${verified ? 'var(--green)' : 'var(--red)'};
             line-height:80px;font-size:32px;color:${verified ? 'var(--green)' : 'var(--red)'}">
          ${verified ? '&#x2713;' : '&#x2715;'}
        </div>

        <div style="margin-top:24px;display:grid;grid-template-columns:1fr 1fr 1fr;
                    gap:12px;max-width:360px;margin-inline:auto">
          <div class="face-metric-item">
            <div class="face-metric-label">Similarity</div>
            <div class="face-metric-val" style="font-size:20px;color:${verified?'var(--green)':'var(--red)'}">
              ${sim.toFixed(4)}</div>
          </div>
          <div class="face-metric-item">
            <div class="face-metric-label">Threshold</div>
            <div class="face-metric-val" style="font-size:20px">${r.threshold}</div>
          </div>
          <div class="face-metric-item">
            <div class="face-metric-label">Person</div>
            <div style="font-size:12px;font-weight:600;color:var(--blue);margin-top:4px;
                        font-family:monospace">${pid}</div>
          </div>
        </div>
        <div style="margin-top:16px;max-width:360px;margin-inline:auto">${fmt.simBar(sim, r.threshold)}</div>
      </div>`;
  } catch(e) {
    Toast.error(e.message);
    resEl.innerHTML = `<p style="color:var(--red)">${e.message}</p>`;
  }
}

File: Deploy/frontend/pages/face_search.py
"""
Face search and verification page.
"""

from __future__ import annotations

import sys
from pathlib import Path

DEPLOY_ROOT = Path(__file__).resolve().parent.parent.parent
if str(DEPLOY_ROOT) not in sys.path:
    sys.path.insert(0, str(DEPLOY_ROOT))

import io
import requests
import pandas as pd
import streamlit as st
from PIL import Image

from backend.config import BACKEND_URL

API = BACKEND_URL


def show():
    st.title("🔍 Tìm kiếm / Xác minh khuôn mặt")

    tab_search, tab_verify = st.tabs(["🔎 Tìm kiếm (1-to-N)", "✅ Xác minh (1-to-1)"])

    # ── Search tab ─────────────────────────────────────────────────────────
    with tab_search:
        st.subheader("Tìm kiếm khuôn mặt – top-K người giống nhất")

        col_left, col_right = st.columns([1, 1])

        with col_left:
            query_file = st.file_uploader(
                "Upload ảnh query",
                type=["jpg", "jpeg", "png", "bmp", "webp"],
                key="search_file",
            )
            top_k = st.slider("Top K", 1, 10, 5)

        if query_file:
            pil = Image.open(query_file)
            col_left.image(pil, caption="Query image", width="stretch")
            query_file.seek(0)

            with st.spinner("Đang tìm kiếm ..."):
                try:
                    r = requests.post(
                        f"{API}/face/search?top_k={top_k}",
                        files={"image": (query_file.name, query_file.read(), query_file.type)},
                        timeout=30,
                    )
                    if r.status_code == 200:
                        data = r.json()
                        results = data.get("results", [])

                        with col_right:
                            st.markdown(f"**{data.get('num_faces', 0)} mặt detect được**")

                            if results:
                                df = pd.DataFrame(results)
                                display = ["person_id", "full_name", "student_code", "class_name", "status", "similarity"]
                                df_show = df[[c for c in display if c in df.columns]].copy()
                                df_show["similarity"] = df_show["similarity"].round(4)

                                best = results[0]
                                threshold = 0.35
                                if best["similarity"] >= threshold:
                                    st.success(
                                        f"✅ MATCH: **{best.get('full_name', best['person_id'])}** "
                                        f"| {best['person_id']} | sim={best['similarity']:.4f}"
                                    )
                                else:
                                    st.warning(
                                        f"⚠️ UNKNOWN. Best={best['person_id']} sim={best['similarity']:.4f}"
                                    )

                                st.dataframe(df_show, use_container_width=True)
                            else:
                                st.warning("Không tìm thấy kết quả.")
                    else:
                        st.error(f"Lỗi {r.status_code}: {r.text}")
                except Exception as e:
                    st.error(f"Lỗi: {e}")

    # ── Verify tab ─────────────────────────────────────────────────────────
    with tab_verify:
        st.subheader("Xác minh 1-to-1 với một người cụ thể")

        try:
            people = requests.get(f"{API}/people", timeout=5).json()
        except Exception:
            people = []

        if not people:
            st.warning("Chưa có người trong database.")
            return

        opts = [f"{p['person_id']} – {p.get('full_name','')}" for p in people]
        chosen = st.selectbox("Chọn người cần xác minh", opts, key="verify_person")
        chosen_id = chosen.split(" – ")[0]

        verify_file = st.file_uploader(
            "Upload ảnh query",
            type=["jpg", "jpeg", "png", "bmp", "webp"],
            key="verify_file",
        )

        if verify_file:
            col_v1, col_v2 = st.columns([1, 1])
            col_v1.image(Image.open(verify_file), caption="Query image", width="stretch")
            verify_file.seek(0)

            if st.button("🔐 Xác minh"):
                with st.spinner("Đang xác minh ..."):
                    try:
                        r = requests.post(
                            f"{API}/face/verify",
                            data={"person_id": chosen_id},
                            files={"image": (verify_file.name, verify_file.read(), verify_file.type)},
                            timeout=30,
                        )
                        if r.status_code == 200:
                            result = r.json()
                            with col_v2:
                                if result.get("is_verified"):
                                    st.success(
                                        f"✅ VERIFIED\n\n"
                                        f"similarity = **{result['similarity']:.4f}**\n\n"
                                        f"threshold = {result['threshold']}"
                                    )
                                else:
                                    st.error(
                                        f"❌ NOT VERIFIED\n\n"
                                        f"similarity = **{result['similarity']:.4f}**\n\n"
                                        f"threshold = {result['threshold']}"
                                    )
                        else:
                            st.error(f"Lỗi {r.status_code}: {r.text}")
                    except Exception as e:
                        st.error(f"Lỗi: {e}")
